### DISCUSSION TITLE
Any good ways to handle object contains option<T> on updating?

### DISCUSSION THREAD

**Discussion by Dennis-Zhang-SH** [Brendonovich/prisma-client-rust#288] [Q&A] [answered]
```rust
pub struct UpdateUserInfo {
    pub address: Option<String>,
    pub role_id: Option<i32>,
}

pub async fn update_user(
    Path(id): Path<i32>,
    c: JwtClaims,
    Json(payload): Json<UpdateUserInfo>,
) -> ... {
        ...
        let mut v: Vec<db::user::SetParam> = vec![];
        if let Some(x) = payload.address{
            v.push(db::user::address::set(x));
        }
        ... // for each filed
        client
                .user()
                .update(db::user::id::equals(id), v)
                .select(user_out::select())
                .exec()
                .await?
}
```

Is there a better solution than this? I am working on writing a proc_macro_attribute also.

**Comment by Brendonovich** [comment]
Personally I like this approach:

```rust
pub struct UpdateUserInfo {
	pub address: Option<String>,
	pub role_id: Option<i32>,
}

let payload: UpdateUserInfo = ...;

let values = [
	payload.address.map(db::user::address::set)	
].into_iter().flatten().collect();

client
	.user()
	.update(db::user::id::equals(id), v)
	.select(user_out::select())
	.exec()
	.await?
```

I've also thought about giving each module a `partial!` macro that can build a struct with certain fields, it'd just be tricky supporting arbitrary attributes and derives.

**Reply by Dennis-Zhang-SH** [reply]
> Personally I like this approach:
> 
> ```rust
> pub struct UpdateUserInfo {
> 	pub address: Option<String>,
> 	pub role_id: Option<i32>,
> }
> 
> let payload: UpdateUserInfo = ...;
> 
> let values = [
> 	payload.address.map(db::user::address::set)	
> ].into_iter().flatten().collect();
> 
> client
> 	.user()
> 	.update(db::user::id::equals(id), v)
> 	.select(user_out::select())
> 	.exec()
> 	.await?
> ```
> 
> I've also thought about giving each module a `partial!` macro that can build a struct with certain fields, it'd just be tricky supporting arbitrary attributes and derives.

this one do looks better, but with all the update apis need to do so for each specific struct(some of them contain a lot fields), I am getting headache. But indeed, implementing such macro would be too tricky in a orm library, I would like to try to implement something close to `bevy_reflect` does.

**Reply by Brendonovich** [reply]
Making something basic that only supports scalar fields wouldn't be too hard and I could probably do it in the next update, it's just a question of customisation and how to handle relations.

```rust
user::partial!(UpdateUserInfo {
    address
    role_id
});

db.user()
    .update(user::id::equals(id), data.to_params())
    .exec()
    .await?
```

**Reply by Dennis-Zhang-SH** [reply]
Exactly, you'll probably need the `db::user` to fulfill as `db::user::field_name::set`, besides declare macros really mess up with the lsp😂, so personally I would prefer doing something like `#[IterObject(db::user)]` on top of the struct, it impl a trait that may look like:
```rust
trait IterObject {
    fn to_params<T: SetParam>() -> Vec<T>;
}
```

**Reply by Brendonovich** [reply]
Seeing as `include!` and `select!` already exist I'm fine with adding `partial!`, `#[IterObject]` on the other hand is a bit dicier. Wouldn't it only solve half of the problem? You'd still have to declare all the struct fields with their `Option<...>` values.

**Reply by Dennis-Zhang-SH** [reply]
> Seeing as `include!` and `select!` already exist I'm fine with adding `partial!`, `#[IterObject]` on the other hand is a bit dicier. Wouldn't it only solve half of the problem? You'd still have to declare all the struct fields with their `Option<...>` values.

I think you could iterate the struct fields in the proc macro, that's at least what `bevy_reflect` did, declare macro still needs to write all the field name, which I think could be skipped, but it its more consistent with other macros in the crate.

**Reply by Brendonovich** [reply]
I'm still not quite sure what you mean, would you mind making a small example with the syntax you're thinking of?

**Reply by Dennis-Zhang-SH** [reply]
> I'm still not quite sure what you mean, would you mind making a small example with the syntax you're thinking of?

Hi this is the example: https://github.com/Dennis-Zhang-SH/iter-object
I am not really familiar with macros so it took me some time.

**Reply by Dennis-Zhang-SH** [reply]
Oh just find out that `set` fn of some fields requires `Option<T>`, others requires `T`, that's a tricky one...

**Reply by Brendonovich** [reply]
> Oh just find out that set fn of some fields requires Option<T>, others requires T, that's a tricky one...

Yeah with the derive approach it wouldn't be uncommon to require `Option<Option<T>>` for some fields.

I managed to get this working and I think I prefer it. It can inherit existing serde and specta functionality, and doesn't require knowing the exact type for each field. It doesn't exactly allow for adding _extra_ fields which would be annoying for something like `rspc` (but that could be done with some custom syntax), but it's more inline with what already exists and is imo easier to implement.
```rust
user::partial!(struct UserPartialType {
    name
    email
});

#[tokio::test]
async fn scalars() -> TestResult {
    let client = client().await;

    let user = client
        .user()
        .create("Brendan".to_string(), vec![])
        .exec()
        .await?;

    let updates = UserPartialType {
        // no update
        name: None,
        // update optional field with Option<Option<T>>
        email: Some(Some("brendonovich@outlook.com".to_string())),
    };

    let updated_user = client
        .user()
        .update(user::id::equals(user.id), updates.to_params())
        .exec()
        .await?;

    assert_eq!(
        updated_user.email,
        Some("brendonovich@outlook.com".to_string())
    );

    cleanup(client).await
}
```

**Reply by Dennis-Zhang-SH** [reply]
Okay, but why do you need a `Option<Option<T>>`, does that mean all my update struct needs to be written in a way like:
```rust
pub struct UpdateInfo {
    pub name: Option<Option<T>>
}
```
I wonder the macro implementation.

**Reply by Brendonovich** [reply]
Say `User` is this model:

```prisma
model User {
    id             String     @id @default(cuid())
    name           String
    email          String?    @unique
}
```

The corresponding `SetParam` enum is something like this:

```rust
enum SetParam {
  Id(String),
  Name(String),
  Email(Option<String>),
}
```

And when we use `partial!` it makes each field you specify optional:

```rust
user::partial!(struct UserPartial {
  name
  email
});

// generates

struct UserPartial {
  name: Option<String>,
  email: Option<Option<String>>
}
```

So `Option<Option<T>>` wouldn't always be necessary, only for optional fields. You then get into the issue of handling them properly with serde (which PCR can do automatically), and it's just annoying to type. May as well tell PCR what fields you want and let it generate the struct itself.

**Reply by Dennis-Zhang-SH** [reply]
Oh, are you going to reform the Set function into set(Option<T>), so that Option<Option<T>> could be flatten and used in the set fn?

**Reply by Brendonovich** [reply]
No need, we can use the `map ... flatten` trick I shared earlier to implement `to_params`:

```rust
impl UserPartial {
  pub fn to_params(self) -> Vec<SetParam> {
    [
      self.name.map(user::name::set),
      self.email.map(user::email::set),
    ].into_iter().flatten().collect()
  }
}
```

Since each field of `UserPartial` is the original field but wrapped in `Option`, we can just `map` everything and the types all work out.

**Comment by Brendonovich** [comment, marked as answer]
`model::partial!` is now available on `main` if you want to try it out!

**Reply by Dennis-Zhang-SH** [reply]
> `model::partial!` is now available on `main` if you want to try it out!

Hi I change my rev to the latest, and got a lot of type annotation errors:
![image](https://user-images.githubusercontent.com/51769674/227400819-b0c26cd2-21a5-43a4-993b-d95bc9acc1e8.png)

**Reply by Brendonovich** [reply]
That's odd - are you on rev `1699cff77f47acbee5d8adf3684d64a15e964ae7`?
Also, could you share your schema for some of the models that are affected by those errors? I'm guessing it's something to do with required fields and `create_many`.

**Reply by Dennis-Zhang-SH** [reply]
> That's odd - are you on rev `1699cff77f47acbee5d8adf3684d64a15e964ae7`? Also, could you share your schema for some of the models that are affected by those errors? I'm guessing it's something to do with required fields and `create_many`.

https://github.com/Dennis-Zhang-SH/eam-backend-rust
I made my repo public, you can check the model at `prisma/schema.prisma`

**Reply by Brendonovich** [reply]
I think your issue is that you have mismatching `prisma-client-rust` and `prisma-client-rust-cli` versions in `Cargo.toml` and `prisma-cli/Cargo.toml`.
As an aside, you don't need to depend on `prisma-client-rust` in `prisma-cli`.

**Reply by Brendonovich** [reply]
I've opened https://github.com/Dennis-Zhang-SH/eam-backend-rust/pull/1 which should make your project build properly.

**Reply by Dennis-Zhang-SH** [reply]
> I've opened [Dennis-Zhang-SH/eam-backend-rust#1](https://github.com/Dennis-Zhang-SH/eam-backend-rust/pull/1) which should make your project build properly.

Ah ha! I totally forgot it, thanks for the helping! But I find out that the struct created by macro is not able to derive `serde::{Serialize, Deserialize}`, I guess manually impl these traits wouldn't be a good choice? maybe can do something like:
```rust
model::partial!{
    UpdateUserInfo {
        name
    },
    Serialize,
    Deserialize
}
``` 
what do you think?

**Reply by Brendonovich** [reply]
Oh whoops I think I forgot to add `Serialize`, `Deserialize` and `Type`. Gimme a bit.

**Reply by Brendonovich** [reply]
Try `rev = "49157998fcdcbe6aa45088bb4f2dd08011393209"`, I've implemented `Deserialize` but probably won't implement `Serialize` since I don't see a need for it

**Reply by Dennis-Zhang-SH** [reply]
> Try `rev = "49157998fcdcbe6aa45088bb4f2dd08011393209"`, I've implemented `Deserialize` but probably won't implement `Serialize` since I don't see a need for it

Hmm, I clean the target and cargo cache, then delete all the registry in ~/.cargo/registry, but I still got the error:
```
the trait `for<'de> db::_::_serde::Deserialize<'de>` is not implemented for `UpdateUserInfo`
```
then I expand the macro, it seems deserialize is not implemented for the struct
![image](https://user-images.githubusercontent.com/51769674/227447813-73d69c72-69a3-490c-960d-d4036845f87b.png)

I checked your code, you were hand writing the deserialize for the struct, wouldn't derive simply work?

**Reply by Brendonovich** [reply]
Hmm, have you updated both `prisma-client-rust` and `prisma-client-cli` and regenerated the client? The implementation code is definitely there:
<img width="599" alt="image" src="https://user-images.githubusercontent.com/14191578/227455887-28735f0d-e3cf-48b8-9d9a-7b00c80eb6e5.png">

As for why I don't use the derive macros - at the moment I implement all macros using `macro_rules`, which doesn't give enough flexibility to customise implementations of `Serialize` and `Deserialize` as much as I'd like. I'm looking at using proc macros instead though (#250) which would allow the derive macros to be used.

**Reply by Dennis-Zhang-SH** [reply]
> Hmm, have you updated both `prisma-client-rust` and `prisma-client-cli` and regenerated the client? The implementation code is definitely there: <img alt="image" width="599" src="https://user-images.githubusercontent.com/14191578/227455887-28735f0d-e3cf-48b8-9d9a-7b00c80eb6e5.png">
> 
> As for why I don't use the derive macros - at the moment I implement all macros using `macro_rules`, which doesn't give enough flexibility to customise implementations of `Serialize` and `Deserialize` as much as I'd like. I'm looking at using proc macros instead though (#250) which would allow the derive macros to be used.

I forgot to regenerate...these little mistakes are killing, sorry for wasting your time.
But ya, a proc macro would actually be nicer, it's more readable and harder to mess up, if you got time to do it.

**Reply by Brendonovich** [reply]
Yeah I'm gonna do it at some point, I'm just sad that they may eliminate the small amount of autocomplete that `macro_rules` allows for. I guess it's ok since there'd be much better error reporting.

**Reply by Brendonovich** [reply]
@Dennis-Zhang-SH Does everything still work if you use `rev = "45cbccf2932f02a6a5219823f6f77cd579b9c741"`? I'm trying out some proc macros.

**Reply by Dennis-Zhang-SH** [reply]
> @Dennis-Zhang-SH Does everything still work if you use `rev = "45cbccf2932f02a6a5219823f6f77cd579b9c741"`? I'm trying out some proc macros.

Sadly no, the macro does not work anymore, it doesn't generate anything.

**Reply by Brendonovich** [reply]
Hmm... the tests are passing - can you push your changes?

**Reply by Dennis-Zhang-SH** [reply]
> Comment options

Sure, I'll push it, FYI, the macro doesn't seems to be able working on relation fields.

**Reply by Brendonovich** [reply]
Yeah it's not supposed to support relations yet since PCR doesn't support nested writes

**Reply by Dennis-Zhang-SH** [reply]
> Yeah it's not supposed to support relations yet since PCR doesn't support nested writes

Okay, you can test on my repo now.

**Reply by Brendonovich** [reply]
it looks like you used the wrong `rev`, but there were issues on my end too. `c7e2368cb624e1647853e14617afaf32c14365b7` should do the trick.

**Reply by Dennis-Zhang-SH** [reply]
> it looks like you used the wrong `rev`, but there were issues on my end too. `c7e2368cb624e1647853e14617afaf32c14365b7` should do the trick.

I changed it locally and it didn't work, so I changed it back, `c7e2368cb624e1647853e14617afaf32c14365b7` works fine.

**Reply by Dennis-Zhang-SH** [reply]
> it looks like you used the wrong `rev`, but there were issues on my end too. `c7e2368cb624e1647853e14617afaf32c14365b7` should do the trick.

Hey I checked your repo in your homepage, I am really interested in spacedrive, is there any way I can contribute? I am a fullstack developer spent most time on devops and rust, I saw you guys are hiring Senior Front-end developer, well I am not  that good at frontend but I think I can learn about it. Currently I can do part-time job, you can assign tasks to me, no salary is needed, but will need it if  turn into full time job:).
